qa/rpc-tests/addressindex.py

```diff
@@ -0,0 +1,363 @@
+
+        # Multiple address arguments, results are the sum
+        check_balance(1, [addr_p2sh, addr_p2pkh], 105 * 12.5 * COIN)
+
+        assert_equal(len(self.nodes[1].getaddresstxids(addr_p2pkh)), 105)
+        assert_equal(len(self.nodes[1].getaddresstxids(addr_p2sh)), 105)
+
+        # only the oldest 5 transactions are in the unspent list,
+        # dup addresses are ignored
+        height_txids = getaddresstxids(1, [addr_p2pkh, addr_p2pkh], 1, 5)
+        assert_equal(sorted(height_txids), sorted(unspent_txids))
+
+        height_txids = getaddresstxids(1, [addr_p2sh], 1, 5)
+        assert_equal(sorted(height_txids), sorted(unspent_txids))
+
+        # each txid should appear only once
+        height_txids = getaddresstxids(1, [addr_p2pkh, addr_p2sh], 1, 5)
+        assert_equal(sorted(height_txids), sorted(unspent_txids))
+
+        # do some transfers, make sure balances are good
+        txids_a1 = []
+        addr1 = self.nodes[1].getnewaddress()
+        expected = 0
+        expected_deltas = [] # for checking getaddressdeltas (below)
+        for i in range(5):
+            # first transaction happens at height 105, mined in block 106
+            txid = self.nodes[0].sendtoaddress(addr1, i + 1)
+            txids_a1.append(txid)
+            self.nodes[0].generate(1)
+            self.sync_all()
+            expected += i + 1
+            expected_deltas.append({
+                'height': 106 + i,
+                'satoshis': (i + 1) * COIN,
+                'txid': txid,
+            })
+        check_balance(1, addr1, expected * COIN)
+        assert_equal(sorted(self.nodes[0].getaddresstxids(addr1)), sorted(txids_a1))
+        assert_equal(sorted(self.nodes[1].getaddresstxids(addr1)), sorted(txids_a1))
+
+        # Restart all nodes to ensure indices are saved to disk and recovered
+        stop_nodes(self.nodes)
+        wait_bitcoinds()
+        self.setup_network()
+
+        bal = self.nodes[1].getaddressbalance(addr1)
+        assert_equal(bal['balance'], expected * COIN)
+        assert_equal(bal['received'], expected * COIN)
+        assert_equal(sorted(self.nodes[0].getaddresstxids(addr1)), sorted(txids_a1))
+        assert_equal(sorted(self.nodes[1].getaddresstxids(addr1)), sorted(txids_a1))
+
+        # Send 3 from addr1, but -- subtlety alert! -- addr1 at this
+        # time has 4 UTXOs, with values 1, 2, 3, 4. Sending value 3 requires
+        # using up the value 4 UTXO, because of the tx fee
+        # (the 3 UTXO isn't quite large enough).
+        #
+        # The txid from sending *from* addr1 is also added to the list of
+        # txids associated with that address (test will verify below).
+
+        addr2 = self.nodes[2].getnewaddress()
+        txid = self.nodes[1].sendtoaddress(addr2, 3)
+        self.sync_all()
+
+        # the one tx in the mempool refers to addresses addr1 and addr2
+        mempool = self.nodes[0].getaddressmempool({'addresses': [addr2, addr1, addr2]})
```

Comment:
Is it significant to the test that `addr2` is duplicated?

Reply:
Yes, this was intentional, I wanted to test that the results are returned in the order corresponding to the arguments, including duplicate arguments. I will add a comment to clarify.